File: problem2.cpp
// Time Complexity : O(V + E)
// Space Complexity : O(V + E)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

// this problem is having dependencies, so clearly, this a graph problem.
// we create adjacency list using map with key as independent items and values as dependent items.
// we  need to find acourse to start with which is completey indpendent of other courses. so we calculate an indegree array 
// to keep track of all course dependencies. 
// we start doing BFS using a queue by adding courses which have indegree values as zero  and then we reduce the values of
//  its dependent courses and proceed further by  adding remaining courses if their indegree becomes zero.
//  we keep track of number of items added to queue and after BFS ,check if count == Number of courses and return answer based on that.

class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        if(prerequisites.empty() || prerequisites.size()==0) return true;
        vector<int>indegree(numCourses,0);
        unordered_map<int,vector<int>>map;
        for(auto p: prerequisites)
        {
            indegree[p[0]]++;
            if(map.find(p[1])==map.end())
            {
                vector<int>v;
                map.insert({p[1],v});
            }
            map[p[1]].push_back(p[0]);
        }
        queue<int>q;
        int cnt = 0;
        for(int i = 0;i<numCourses;i++)
        {
            if(indegree[i]==0){
                q.push(i);
                cnt++;
            }
        }
        while(!q.empty())
        {
            int curr = q.front();q.pop();
            for(int i = 0;i<map[curr].size();i++)
            {
                indegree[map[curr][i]]--;
                if(indegree[map[curr][i]] ==0){
                    q.push(map[curr][i]);
                    cnt++;
                }
            }
        }
        if(cnt == numCourses) return true;
        return false;
    }
};

File: Problem - 2.java
// Time Complexity : O(v+E)
// Space Complexity :O(v+E)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : Algorithm was tough


// Your code here along with comments explaining your approach

/*
We will create an in degree array of the node, which represents incoming edges to that vertices and in general that would represent the dependency

Once we create the in degree, we will pick up the independant vertice i.e the ones that has no inwards ege i.e 0 count nodes and will reduce the count by one of the vertices that depends on them and then this is how loop condtinues

*/
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
		
		if(prerequisites.length == 0) return true;  // no prereqs than we can take any number of courses

		// We will convert the edges into adjacency list
		HashMap<Integer,List<Integer>> list = new HashMap<Integer,List<Integer>>();
		
		HashMap<Integer,Integer> indegree = new HashMap<Integer,Integer>(); // we could have taken an array as well


		// Creating the adjacency list
		for(int i=0;i<prerequisites.length;i++){ // O(E)
			// the 1 position vertex is independant
			// the 0 position vertex in dependent
			if(list.get(prerequisites[i][1]) == null){
				list.put(prerequisites[i][1], new ArrayList<Integer>());
			}
			list.get(prerequisites[i][1]).add(prerequisites[i][0]); // we will store that how many vertices are dependant on it
			
			//updating the indegree of dependant
			if(indegree.get(prerequisites[i][0]) == null){
				indegree.put(prerequisites[i][0],0);
			}

			// Since the vertex has a dependency so increasing its count
			indegree.put(prerequisites[i][0],indegree.get(prerequisites[i][0]) +1 );
		}

		

		Queue<Integer> queue = new LinkedList<Integer>();

		int courses = 0; // number of course we can take

		// we will go over the indegree and see whos dependency count is 0
		// we will add all the independant vertex  to the queue
		// Important , the numCourses are the total numCourses, which means the edges would be from 0 to numCourses-1
		// there would be vertex (course) who has no edge, so we can count them as independant in this way 
		// if we start from 0 to numCourses
		for(int vertex =0;vertex<numCourses;vertex++) { // O(V)
			if(indegree.get(vertex) == null || indegree.get(vertex)==0){
				queue.add(vertex);
				courses++; // this means that this many course can be taken without any dependency
			}
		}
		if(courses >= numCourses) return true; // if this is the number of courses needs to be taken, then we are done here

		// we will take each independant vertex out of the queue, reduce the count of indegree of vertex that are dependent on this
		// if any vertex indegree count becomes 0 , we will add into the queue and increase the courses count

		while(!queue.isEmpty()){ // O (E)

			int independantVertex = queue.poll();

			
			if(list.get(independantVertex) == null) continue;
			// go to each dependency and decrease the indegree count
			for(int dependent : list.get(independantVertex)){
				// since the independant vertex is being taken out
				// we will reduce the count of all the vertices and nodes which were dependent on this
				int newCount= indegree.get(dependent);
				newCount--;
				indegree.put(dependent,newCount);
				if(newCount == 0) {
					queue.add(dependent); // since now it become independant , we can add it to our queue
				courses++;// increasing the count of number courses can be taken since it has not any dependencies	
				}

			}
				if(courses >= numCourses){
					return true; // if we reached the required number of courses, then no need to check further
				}
		}

			// if after all the processing, we still did'nt reach the number of courses, then return false
			return false;

    }
}